File: template-method/java/DataAnalyzerBase.java
package TemplateMethod;

// Metot iskeletimizi ve soyutlanmış metotları içermektedir.
// UML diyagramındaki AbstractClass yapısına denk gelmektedir.
abstract class DataAnalyzerBase {

    // Metot iskeletimiz.
    public void analyzeDataOfLastWeek() {
        Object data = getDataOfLastWeek();
        Object result = analyzeData(data);
        exportResult(result);
    }

    private Object getDataOfLastWeek() {
        // Burada veri verinin çağrılma süreçleri yer alacaktır.
        // Default olarak bir obje dönülmektedir.
        System.out.println("Method worked: GetDataOfLastWeek");
        return new Object();
    }

    private Object analyzeData(Object data) {
        // Burada verinin analizi süreci yer almaktadır.
        // Analiz sonuçları default olarak bir obje dönmektedir.
        System.out.println("Method worked: AnalyzeData");
        return new Object();
    }

    // Alt sınıflar tarafından uygulanmasını istediğimiz metottur.
    // Metot iskeleti içerisinde farklılık gösterecek olan metotumuzdur.
    public abstract void exportResult(Object result);
}

File: template-method/python/data_analyzer_base.py
from abc import abstractmethod


class DataAnalyzerBase:
    """
    Metot iskeletimizi ve soyutlanmış metotları içermektedir.
    UML diyagramındaki AbstractClass yapısına denk gelmektedir.
    """
    data: object
    result: object

    def analyze_data_of_last_week(self):
        # Metot iskeletimiz.
        self.data = self.get_data_of_last_week()
        self.result = self.analyze_data(self.data)
        self.export_result(self.result)

    def get_data_of_last_week(self):
        # Burada veri verinin çağrılma süreçleri yer alacaktır.
        # Default olarak bir obje dönülmektedir.
        print("Method worked: GetDataOfLastWeek")
        return object()

    def analyze_data(self, data: object):
        # Burada verinin analizi süreci yer almaktadır.
        # Analiz sonuçları default olarak bir obje dönmektedir.
        print("Method worked: AnalyzeData")
        return object()

    # Alt sınıflar tarafından uygulanmasını istediğimiz metottur.
    # Metot iskeleti içerisinde farklılık gösterecek olan metotumuzdur.
    @abstractmethod
    def export_result(self, result: object):
        pass
